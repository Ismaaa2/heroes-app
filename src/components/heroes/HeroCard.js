import React from "react";
import { Link } from "react-router-dom";



export const HeroCard = ({
  id,
  superhero,
  publisher,
  alter_ego,
  first_appearance,
  characters,
  img
}) => {
  return (
    <div className="card ms-3" style={{ width: 300, backgroundColor: '#393c40', borderRadius: 6 }}>
      <div className="row no-gutters">
        <div className="cold-md-4 ">
          <img
            src={ img }
            style={{ height: 300, width: '100%', borderRadius: 6 }}
            className="carg-img"
            alt={alter_ego}
          />
        </div>
        <div className="col-md-8">
          <div className="card-body">
            <h5 className="card-title">{superhero}</h5>
            <p className="card-text"> {alter_ego} </p>
            {   
                (alter_ego  !== characters)
                        && <p className="card-text">{characters}</p>
            }
            <p className="card-text">
                <small className="text-muted">{first_appearance}</small>
            </p>
            <Link to={`./heroe/${id}`} style={{ textDecorationLine: 'none', color: 'maroon' }}>
                    MÃ¡s...
            </Link>
          </div>
        </div>
      </div>
    </div>
  );
};
