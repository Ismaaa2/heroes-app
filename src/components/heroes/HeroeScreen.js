import React, { useMemo } from "react";
import { useParams, Redirect } from "react-router-dom";
import { getHeroesById } from "../../selectors/getHeroesById";


// import batman from '../../assets/heroes/dc-batman.jpg';  estÃ¡tico

const heroeImages = require.context( '../../assets/heroes', true );

export const HeroeScreen = ({ history }) => {

  const { heroeId } = useParams();

  const heroe = useMemo(() => getHeroesById(heroeId), [heroeId])

  if (!heroe) {
    return <Redirect to="/" />;
  }


  const handleReturn = () => {

    if( history.length <= 2 ){
      history.push('/')
    }else{
      history.goBack()
    }
  }

  return (
    <div className="row mt-5">
      <div className="col-4">
        <img
          // src={`/assets/heroes/${heroe.id}.jpg`}  desde assets
          // src={batman} impoort
          src={heroeImages(`./${heroeId}.jpg`).default}
          alt={heroe.alter_ego}
          className="img-thumbnail"
        />
      </div>
      <div className="col-8">
        <h3>{heroe.superhero}</h3>
        <ul className="list-group list-group-flush">
            <li className="list-group-items"><b> Alter ego: </b>{ heroe.alter_ego }</li>
            <li className="list-group-items"><b> Publisher: </b>{ heroe.publisher }</li>
            <li className="list-group-items"><b> First appearance: </b>{ heroe.first_appearance }</li>
        </ul>

        <h5>Characters</h5>
        <p>{heroe.characters}</p>

        <button className="btn btn-danger" onClick={handleReturn}>
          Return
        </button>
      </div>
    </div>
  );
};
