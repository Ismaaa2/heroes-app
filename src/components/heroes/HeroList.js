import React, { useMemo } from 'react'
import { getHeroesByPublisher } from '../../selectors/getHeroesByPublisher'
import { HeroCard } from './HeroCard';

const heroeImages = require.context( '../../assets/heroes', true );


export const HeroList = ({ publisher }) => {

    const heroes = useMemo(() => getHeroesByPublisher( publisher ), [publisher])

    return (
        <div style={{ display: 'flex', flexWrap: 'wrap', gap: 45}}>
           {
               heroes.map(hero => {
                    const img = heroeImages(`./${hero.id}.jpg`).default
                    return <HeroCard key={hero.id} { ...hero } img={img} />
               }
               )
           } 
        </div>
    )
}
