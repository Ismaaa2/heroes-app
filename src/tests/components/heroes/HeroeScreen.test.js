import { mount } from "enzyme";
import { MemoryRouter, Route } from "react-router";
import { HeroeScreen } from "../../../components/heroes/HeroeScreen";

describe("<HeroeScreen />", () => {
  const history = {
    length: 10,
    push: jest.fn(),
    goBack: jest.fn(),
  };

  test("debe mostrar el componente redirect si no hay argumentos en la URL", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/heroe"]}>
        <HeroeScreen history={history} />
      </MemoryRouter>
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find("Redirect").exists()).toBe(true);
  });

  test("debe de mostrar un heroe si el parametro existe y si se encuentra", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/heroe/marvel-spider"]}>
        <Route path="/heroe/:heroeId" component={HeroeScreen} />
      </MemoryRouter>
    );

    expect(wrapper.find("h3").exists()).toBe(true);
  });

  test("debe de regresar a la pantalla anterior con push", () => {
    const history = {
      length: 1,
      push: jest.fn(),
      goBack: jest.fn(),
    };

    const wrapper = mount(
      <MemoryRouter initialEntries={["/heroe/marvel-spider"]}>
        <Route
          path="/heroe/:heroeId"
          component={() => <HeroeScreen history={history} />}
        />
      </MemoryRouter>
    );

    wrapper.find("button").prop("onClick")();

    expect(history.push).toHaveBeenCalledTimes(1);
    expect(history.push).toHaveBeenCalledWith("/");
  });

  test("debe de regresar a la pantalla anterior con goBack", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/heroe/marvel-spider"]}>
        <Route
          path="/heroe/:heroeId"
          component={() => <HeroeScreen history={history} />}
        />
      </MemoryRouter>
    );

    wrapper.find("button").prop("onClick")();

    expect(history.goBack).toHaveBeenCalledTimes(1);
  });

  test("debe de llamar al redirect si el hero no existe", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/heroe/marvel-spider232"]}>
        <Route
          path="/heroe/:heroeId"
          component={() => <HeroeScreen history={history} />}
        />
      </MemoryRouter>
    );

    expect(wrapper.text()).toBe('');
  });
});
