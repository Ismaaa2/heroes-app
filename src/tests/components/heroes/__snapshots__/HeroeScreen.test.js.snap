// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HeroeScreen /> debe mostrar el componente redirect si no hay argumentos en la URL 1`] = `""`;
